OurUmbraco/Project/Api/ProjectUploadController.cs

```diff
@@ -0,0 +1,111 @@
+﻿using Newtonsoft.Json;
+using OurUmbraco.Wiki.BusinessLogic;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Net;
+using System.Net.Http;
+using System.Threading.Tasks;
+using System.Web;
+using System.Web.Http;
+
+namespace OurUmbraco.Project.Api
+{
+    public class ProjectUploadController : UmbracoMemberAuthApiController
+    {
+        /// <summary>
+        /// Used by the package CLI tool, gets a list of project files to check if one already exists with the same name
+        /// </summary>
+        public IEnumerable<WikiFile> GetProjectFiles()
+        {
+            // The project/member id are exposed as base properties and are resolved from the identity created on the request
+            var files = WikiFile.CurrentFiles(ProjectNodeId);
+            return files.Where(x => x.FileType == "package" || x.FileType == "hotfix").OrderByDescending(x => x.Version.Version).ToList();
+        }
+
+        /// <summary>
+        /// Posts a package file + some related meta data to update a package from the CLI package tool.
+        /// Includes the zip, whether it should be current, the compatible .NET and Umbraco versions.
+        /// </summary>
+        /// <returns>HttpResponseMessage</returns>
+        /// <exception cref="HttpResponseException"></exception>
+        [HttpPost]
+        public async Task<HttpResponseMessage> UpdatePackage()
+        {
+            // Check if the request contains multipart/form-data.
+            if (!Request.Content.IsMimeMultipartContent())
+            {
+                throw new HttpResponseException(HttpStatusCode.UnsupportedMediaType);
+            }
+
+            string root = HttpContext.Current.Server.MapPath("~/App_Data");
+            var provider = new MultipartFormDataStreamProvider(root);
+
+            try
+            {
+                // Read the form data.
+                await Request.Content.ReadAsMultipartAsync(provider);
+
+                // verify that a file exist (and only one)
+                if (provider.FileData.Count == 0)
+                {
+                    return Request.CreateErrorResponse(HttpStatusCode.InternalServerError, "No file attached");
+                }
+                else if (provider.FileData.Count > 1)
+                {
+                    return Request.CreateErrorResponse(HttpStatusCode.InternalServerError, "You can only upload one file");
+                }
+                else
+                {
+                    var packageFile = provider.FileData.FirstOrDefault();
+                    var fileName = packageFile.Headers.ContentDisposition.FileName.Replace("\"", "");
```

Comment:
Hmm, need to make really sure that we can't do path traversal here.. if the filename ends in something like `../../web.config` maybe that would not be great? Not sure if this is a real vulnerability but need to look into that.

Reply:
As discussed on Slack, this seems to not be a problem anyways 🙂